Exclusive image bounds (max)

HughMacdonald opened the pull request:
Not ready to be merged yet, but I'm putting in this PR for discussion as I go through it all.

This is the lowest level of modifications for converting the internal image bounds to be exclusive at the top end.

This just contains changes to Format and ImagePlug (and their respective tests)

All tests pass on these two classes.

It's going to be hard to confirm that they do full work until ImageReader and either ImageView or ImageWriter have been modified accordingly.

HughMacdonald commented:
Rebased with Format.inl fix.
Please ignore the current ImageView commit.

johnhaddon commented:
Nice one - thanks for taking care of that so quickly. If you get a chance as you're going through this, please feel free to add additional unit tests as you see fit to give us extra confidence in the correctness of the changes...

johnhaddon commented:
One other thing - let's update the class documentation for ImagePlug to unambiguously document the new convention, and point out the differences to OpenEXR/Cortex.

HughMacdonald commented:
Will do.

So current state is:

ImageView has some issues - when creating a 50x50 Constant, ImageView has some issues around the edges. I've hit a bit of a bump in that IECore::ImagePrimitive requires inclusive bounds, but I'm dealing with that in ImagePlug::image() (and segfaulting somewhere along the way, but I'll track that one done)

I feel like this PR is going to have a lot of re-basing, as I may find issues with lower level classes as I'm working on higher level ones. Are you okay with losing exact diffs in order to keep the commits clean? Alternatively, I can not do any rebasing until the very end of the process.

johnhaddon commented:
I'm OK with whatever process keeps you sane and gives us a nice clean PR for final review. If you're working with the ImageView, you might find it useful to use this temp fix for the mipmap filtering so you can see what's going on a bit better :

https://github.com/johnhaddon/gaffer/commit/ba109de713a047246ea2fcc799f66c19049f43e6

HughMacdonald commented:
I think I'm going to not do any re-basing until the end then - it'll be much easier to keep track of then!

Thanks for that update - I'll apply that

HughMacdonald commented:
Now including ImageView, ImageReader and ImageWriter

HughMacdonald commented:
FYI : Currently, the tests that don't pass are:

Display (seg faults)
ImageTransform
Reformat
Sampler
ImageStats
ObjectToImage

I'm looking into why DisplayTest seg faults first, then will work through the others.

johnhaddon commented:
Nice progress Hugh. Is it alright with you if I wait till you have everything passing before I do a full review?

HughMacdonald commented:
Absolutely fine - there's some tweaking that I'd like to do in places (there are a number of y-up/y-down switches that aren't using the code in Format, and they've really caught me out a few times - I'll aim to do some static inline methods in Format that can be used for times when creating a Format object isn't necessary.

HughMacdonald commented:
Here's a slightly amusing one...

OpenEXR, as we know, deals with inclusive image bounds.

However, Box2i is set up to be exclusive..

A Box2i from (0,0)-(0,0), which defines a 1-pixel area, returns "False" for hasVolume().

If you put a Crop node into Gaffer, and crop it down to be only 1 pixel wide or high, and thus still giving image data, you'll find that the data window overlay disappears, as ImageView is using hasVolume() to decide whether to draw it or not.

The issue has come up because ImagePlug::image(), if there is no data, returns a 1x1 black pixel. 

If ImageView is given a completely empty image, creating the texture to draw fails. I'm going to look at doing a check at that stage, and only drawing the texture if pixel data actually exists in it.

johnhaddon commented:
I wouldn't worry too much about dealing with parts of the viewer that were broken already as part of this pull request. Let's keep it as focussed as possible on just the change of convention. The ImageView really needs reimplementing to not use ImagePlug::image() at all anyway, for other reasons  - we can address those sorts of edge cases then...

HughMacdonald commented:
Current state:

The following tests seg fault:

ImageWriterTest (this definitely used to work, so I've obviously introduced something. I think that all of these seg faults are the same thing)
DisplayTest
ImageSamplerTest
ObjectToImageTest


The following tests fail:

ImageTransformTest.testBoxFilter
ImageTransformTest.testHermiteFilter
ImageTransformTest.testSincFilter (yes, just these three filters - the others pass)
SamplerTest.testOutOfBoundsSampleModeBlack (I think that this may be the test that needs fixing - will investigate tomorrow)

HughMacdonald commented:
John - I had to fix an issue one way or another - I was getting an error because I was passing an empty image through. Would you rather I get it working with the single pixel that it had before? Currently, the change is this commit: https://github.com/HughMacdonald/gaffer/commit/b2338595d8197f85d7c8f3de4bb2d6111b43eb46

johnhaddon commented:
Hang on, have you adjusted the bounds returned by `ImagePlug::image()` to be exclusive? That's a Cortex image being returned, so we should follow the Cortex/EXR convention within that, however confusing that might be. Or have I missed something?

HughMacdonald commented:
Nope - it shifts them back to being inclusive there, but then sets them back to exclusive in ImageViewGadget.

Both conversions are commented for clarity. (Actually, I lie - the conversion from exclusive to inclusive in ImagePlug::image is commented, the conversion from inclusive to exclusive in ImageViewGadget() is not commented.)

johnhaddon commented:
I'm still confused as to why the "empty image" problem is different now than it was before. But if the change is as minimal as the commit you linked to, then that's fine. I'm really feeling like we need to get this ImageView knocked into shape soon, managing its own threading and texture tiles and avoiding `ImagePlug::image()` entirely.

HughMacdonald commented:
It would be good to get feedback on this commit at some point: https://github.com/HughMacdonald/gaffer/commit/884444db9fc561a94107cdf01c6a54558af39178

This creates some static inline functions in Format for doing the coordinate conversions, requiring windowMinY and windowMaxY values. This allows for them to be used any time a coordinate conversion is required, which may not always have a Format object available.

johnhaddon commented:
I always felt like those existing methods felt a bit unwieldy, but that they were definitely preferable to not having them, since they make the meaning much more clear. I feel the same way about the new ones, except they feel a bit more unwieldy because of the extra arguments. Out of interest, where do we need to do the conversions but without access to a format?

HughMacdonald commented:
It's generally anywhere that we're converting to or from OIIO or ImagePrimitive.

Off the top of my head:

ImageReader::computeChannelData
ImagePlug - CopyTiles::operator()

Oddly enough, not ImageWriter - the Scanline writer seems to be happy writing y-up.

HughMacdonald commented:
To be honest, those areas do work now, so I'm happy to roll back on this one - my main thought was that it would be more straight-forward to understand in future.

Another option would be, whenever we do a conversion, create a temporary Format object which encompasses the area being converted (usually tile by tile) and then use the existing methods.

johnhaddon commented:
Yeah, I was wondering if it might just be clearer to construct a Format object - if that's possible in each of the use cases then I think that'd get my vote.

HughMacdonald commented:
Cool. I'll revert the Format change then, and go that way where it's appropriate.

Thanks for the feedback

HughMacdonald commented:
Latest update:

All tests now pass apart from DisplayTest, which seg faults.

HughMacdonald commented:
I think I'm ready for review here. Travis seems happy that everything passes!

johnhaddon commented:
Hmm. I should have thought of this before, but we have a problem with previously saved files here don't we? Any FormatPlug values saved in old files are in the inclusive form, so when we load those old files, everything ends up a pixel smaller.

Here are a couple of options :

1. Make an intermediate version of Gaffer where we serialise formats in the `Format( width, height )` form if they have an origin of 0,0. By saving a script in that version, you prepare it for loading in the new version.
2. Put a compatibility wrapper around `FormatPlug.setValue()`, and do the appropriate offsetting if it would mean the format would then match one of the registered ones.

Neither of those are foolproof, but given that there is no facility in the UI for creating formats which would defeat either of those approaches, they might be good enough? If that's the case, then maybe 2. is best because it requires nothing of the user?

HughMacdonald commented:
Hmm - yup - that's going to be a slightly tricky one.

I wouldn't be averse to actually doing both 1 and 2 - if a format has 0,0 as its origin, serialising out as Format( width, height, pixelAspect ) could be quite a lot cleaner anyway.

Why don't I take a look at both of these, even if we end up not going that route, it's a new area of Gaffer for me to get my head around.

johnhaddon commented:
Sounds good - thanks. For 1, take a look at src/GafferImageBindings/FormatBinding.cpp. For 2, take a look at the various compatibility config files in startup/GafferScene and startup/GafferImage - these inject python methods to maintain compatibility without needing to complicate the libraries themselves.

johnhaddon commented:
Although the problem with 2 will come when we do open up the UI to allow new sizes to be entered directly - there'll be some special sizes that don't work. Let me do some investigation...

HughMacdonald commented:
How about a new (hidden) plug in Format called exclusiveBoundsPlug. This could default to False, but be set to True for any newly created Formats. That way, if a Format is loaded in with exclusiveBounds set to True, it will skip the check, and only old formats would have it set to False.

Alternatively, a formatVersionPlug, which would default to 1 and be set to 2 for any new Formats created.

HughMacdonald commented:
Actually, I've just had a thought - doing 1 may well help us with 2...

Are we only worried about the default formats created in scripts? So only generally worried about formats  with a (0,0) origin?

If so, then, by also doing 1, we know that any Format that was serialised with a Box2i and a min of (0,0) needs tweaking, because any Format with a min of (0,0) serialised post change will only be serialised as width, height.

HughMacdonald commented:
(sorry for spamming!)

This could cause an issue with the future "create new formats" ability in the GUI, unless that part of the GUI treats new formats with (0,0) min positions as a special case and creates them with the (width, height) constructor...

The only worry then would be with any external python script that creates formats. Old ones would continue to work, but new ones that may allow for custom boxes may be presuming that they're exclusive, but creating using a Box2i, which would be adding 1 to the top end.

Anyway, I've pushed these two tweaks - keeping them as separate commits at the end rather than rebasing for now.

johnhaddon commented:
I'm not super keen on the "exclusiveBoundsPlug" idea - generally our aim with this type of problem is to keep the library as clean as possible, and deal with compatibility in a config file, where we can tolerate a bit more hoop jumping.

I think this is what we're looking for :

https://github.com/ImageEngine/gaffer/issues/1436

If we can read the version information from the file currently being loaded, we can determine whether or not we need to add the offset. Old files don't have the version in them, but if we add that feature to coincide with the Format rollout, we know that files without versions need the fix, and those with don't. We also need to determine if the `setValue()` is happening during loading or not, but it looks like traceback can tell us that.

How does that sound? I could take a look at that on Monday (I'm off tomorrow). What do you reckon @andrewkaufman ?

andrewkaufman commented:
sounds good

HughMacdonald commented:
Yeah - that sounds good to me.

Why would we need to check for setValue() during loading? Wouldn't it make more sense to go the other way around - have the compatibility method check for the script version, and only change it where it's lower than the one where the spec was changed.

Also, I actually made the compatibility function on Format.__init__, as it depends on whether Format is initialised with a Box2i or width, height. Once the Format object exists, this can't be determined any more.

johnhaddon commented:
Well, I'm presuming that we want all uses of the API to be using the new convention - we're not providing backwards compatibility for the API at all. We're providing backwards compatibility for old files, and just want to intercept the point at which the format plug is given a value when loading. In that `setValue()` call, if we determine that the file is an old one, we know we need to pad by 1, if it's not an old one, we can continue as normal (the new normal).

HughMacdonald commented:
The one thing that doing it in FormatPlug will miss, though, is the calls to GafferImage.Format.registerFormat in the old Gaffer scripts, though, won't it?

Whichever way it goes, though, the version will need to be changed to the current one as soon as the script has finished being loaded. So the process would be:

* Trigger script load
* Set version from script
* Load script, allowing compatibility functions to run as and when needed
* Set version to current
* Any new calls no longer make changes

Doing it in Format.\_\_init\_\_ won't cause the changes to happen through the API unless the user has very specifically set the version to an older version.

HughMacdonald commented:
I had another thought on this one (I seem to spend my rides home at the moment pondering Gaffer code... This is maybe not something I should admit to...)

There could also be a global (or script?) isLoadingPlug that is flipped to True when the script load starts, and False when it finishes. I could imagine uses for this beyond what we've been discussing here, and it could be used for exactly this in conjunction with a script version plug.

johnhaddon commented:
I've opened PR #1451 to provide the version information needed to provide backwards compatibility. 

Good point about the format registrations needing adjustment as well Hugh - I had forgotten about those. They're a little tricky actually, because they're not associated with a parent script in the same way that a `setValue()` call is. I think you may need to use `inspect.stack()` to see if we're inside a `ScriptNode.load()` call, and retrieve the script from there, but since we need to only enable the feature during loading we would need to be doing something similar anyway.

Although overriding `__init__()` will cope with both `setValue()` and `registerFormat()` in one place, I'd rather we dealt with them each individually, so that the compatibility hacks have less impact on subsequent uses of the API after loading.

What sort of things did you have in mind for the "isLoading" feature? I've had bad experiences with similar features in other packages in the past, so I've been trying to avoid needing something like that in Gaffer. If possible, I would prefer to have a slightly nasty `inspect.stack()` approach hidden away in a config file than introduce a new feature which potentially has its own problems. Perhaps we can try to do that first, and then continue the discussion about "isLoading" if we come across compelling use cases in the future?

HughMacdonald commented:
Cool - let me look at splitting it up into registerFormat() and setValue() overrides then.

I guess, with the version metadata that you mentioned, knowing that we're loading wouldn't be an issue, if, as I mentioned over on that PR, the version is reset to the current after loading. The compatibility code should never be making changes if the script metadata says the same as the current version, so any FormatPlugs being set after the load has finished will find the version metadata the same (or newer) than the one it's checking against.

HughMacdonald commented:
As discussed in #1451, my code will require knowledge of which version Gaffer will be considered when it is released with this code. For now, I'm leaving a TODO comment in there and presuming 0.16.0.0

HughMacdonald commented:
So, a bit of a hitch... I'm trying to figure it out, but thought I'd ask here too...

I'm in startup/GafferImage/formatCompatibility.py and I need the scriptNode to be able to get the metadata. When overriding FormatPlug.setValue(), I have the plug, so I could just ask for the ancestor, but I don't have the same with Format.registerFormat, as that's a static method.

johnhaddon commented:
Hey Hugh - that's the problem I was referring earlier to when I said "because they're not associated with a parent script in the same way that a setValue() call is". My hope is that you'll be able to use `inspect.stack()` to find out if a load call is happening, and to retrieve the script node - is that the angle you're working from?

HughMacdonald commented:
Interesting thing that came up from this - registered formats are across the whole Gaffer application, not just script by script. I guess that made sense from the fact that they aren't registered on a Script Node.

One thing that this could lead to, especially when Gaffer allows formats to be defined in the GUI, is scripts with more relatively arbitrary formats defined.

HughMacdonald commented:
Ah - got it - I had read something else from that earlier.

Sounds good - I'll work with that.


Cheers

HughMacdonald commented:
Updated to support serialised version info in #1451.

HughMacdonald commented:
I'm going to rebase this with the new master, to ensure that it includes #1451

HughMacdonald commented:
I've just realised that the tests failed. It's FormatTest due to the last few commits. Will fix up now.

HughMacdonald commented:
Actually, hit a bit of an issue here with relation to #1451. I wrote a description of it in there - as it's a closed PR, I guess it could be re-opened?

HughMacdonald commented:
Shifting formatCompatibility conversation over to this PR's conversation.

I considered two options for this, neither of which is pretty. This is because the ScriptNode.load method is in C++, so it gets skipped in the stack.

The stack at the point I'm inspecting it shows where the method is being called, but just the code as written in Python, so the script node could be called anything (it's scriptNode.load() in gui-1.py, script.load() in FileMenu.py).

After the stack frame that shows the call, the next one, which in pure python would be inside ScriptNode.load() is inside the text of the gaffer script (listed as "<string>" in <module>)

The two approaches that I thought of are:

* Assume that "<string>" in <module> means that a script is being loaded. This, however, picks up the test case where s2.execute( s.serialise() ) is called, which we don't want.
* Do a regex against possible calls to ScriptNode.load() ("[^\w]script(Node)?.load\(") and search the line of code being executed. This has the danger of potentially missing a case where it should have caught it.


Any thoughts?

Cheers

johnhaddon commented:
Thanks for looking into this Hugh - it's not pretty as you say. I had assumed that the `load()` call would appear unambiguosly in the stacktrace, but it seems that boost::python functions don't actually get an entry in the stack trace - they seem to be fairly opaque as far as python itself is concerned.

I have one more candidate to throw into the ring - wrapping the load method to track the current state, something like this  :

```
__currentlyLoadingScript = None
def __load( self, *args, **kw ) :

	global __currentlyLoadingScript
	__currentlyLoadingScript = self

	try :
		self.__originalLoad( *args, **kw )
	finally :
		__currentlyLoadingScript = None

Gaffer.ScriptNode.__originalLoad = Gaffer.ScriptNode.load
Gaffer.ScriptNode.load = __load
```

Again, not pretty, but a bit more robust I think, and simpler than examining the stack. Do you think that could work? If this becomes a common occurrence, we can definitely consider making a proper API for getting at this stuff, but I'd like to first see if we can scrape by with just a config file.

HughMacdonald commented:
That would certainly work - are you thinking that that code would go in formatCompatibility.py then?

johnhaddon commented:
Yes, for now it would just go in formatCompatibility.py. I'm reluctant to add public APIs to encourage different behaviours during loading, so for now I'd just like to do it as minimally as possible and see how it pans out. The idea behind the compatibility configs is that we don't mind putting some fairly nasty hacks in them as long as it enables us to keep the libraries themselves clean. I also plan to introduce an environment variable to enable/disable the compatibility, and then have Travis run all tests both with and without. After a period running that way, we intend to strip the compatibility stuff out for a version 1.0.0.0 release...

HughMacdonald commented:
Just pushed that update, and added a test to ensure that formatCompatibility was working properly

HughMacdonald commented:
Just pushed updates - I need to do a merge with the lastest master, so it's not running the tests.

Are you okay with continuing the review based on the current code, and I'll get it merged (it was those ImageWriter changes that are conflicting) and re-pushed.

johnhaddon commented:
Sure thing.

HughMacdonald commented:
So I think I've successfully rebased from master, plus I've tweaked Resample and Resize (and tests) to take the image bounds change into account.

All GafferImageTests pass on my machine, so let's see how Travis feels about them.

johnhaddon commented:
Just a quick note to let you know that although I've been quiet I am chugging through a review on this. I'm tidying a couple of things up along the way and there are a couple of things I still need to look at a bit more closely, but it's basically looking really good - thanks for the epic effort.

johnhaddon commented:
This is as far as I've got today - mostly just little tweaks but a couple of important fixes too :

https://github.com/johnhaddon/gaffer/tree/exclusiveImageBoundsTweaks

The viewer's still not working quite right - try viewing an empty ImageReader or Display node and it displays a bogus data window when the data window is empty. It seems like you have more changes in the viewer than what might be necessary just to account for the change? I have a feeling if I revert all of the viewer changes, then there's just a single change to make to use fromEXRSpace() and we're back to the status quo.

I'm a bit dubious about the test for writing an empty data window as well - shouldn't we be able to write an empty window to an EXR?

Other than that I think we're getting there.

HughMacdonald commented:
Thanks for that - a few oversights in there, and a couple of things I wasn't aware of.

Let me take a look at ImageView and see what I can do. What's the best way to do this with your branch too? Should I continue making changes on my own PR branch?

The writing an empty data window thing is something that was in there before, so I kept that functionality. I'll have a dig and check whether EXR can support a completely empty data window.

johnhaddon commented:
Thanks Hugh - if you could take a look at the ImageView that would be great. I was getting a bit tired at that point yesterday and was finding it hard to see what was going on. I know there are existing bugs in the ImageView that would be nice to address, but I feel that for this particular PR it would be best to limit things to just the matter in hand - adjusting the bounds convention. I might be wrong, but it felt like more was going on in there. It kindof sucks but I'd almost rather keep the bugs in there so I can make a case for a bigger rewrite of the image viewer, with a more modular approach, and managing tiles individually rather than with one unwieldy slow megatexture.

I think I'm right in saying you also added a test case asserting that the empty-data-window-becomes-a-full-data-window behaviour is correct? I wasn't so keen on that because I don't think it is. If the ImageWriter behaviour hasn't actually changed, then I'd probably just leave that new test out and leave it all for a separate PR.

As for how we work, maybe you could pull my changes onto your PR branch, and work on top of those?

johnhaddon commented:
One other thought I had while going through this stuff. It feels great that we've abstracted most of the coordinate system shuffling into the fromEXRSpace/toEXRSpace methods and no longer have to give much thought to what happens internally. But we don't have a way of constructing a Format from an EXR display window initially so it's a slightly leaky abstraction - it seemed that the Display node in particular could benefit from being able to do that. Perhaps a `Format( displayWindow, aspectRatio, bool fromEXRSpace = false )` constructor could help there?

HughMacdonald commented:
I've pulled your changes in to my branch.

Re the empty window test - I was just writing an extra check - that check is also done in testDefaultFormatWrite(), although in that case, it's checking against pre-saved images.

Happy to remove the new test for now. Internal functionality certainly still hasn't changed.

I'll add in that new Format constructor option - sounds good to have that as an option.

HughMacdonald commented:
I've also added in a couple of extra constructor bindings for Format - it's been bugging me a bit that C++ allows creating of a Format with just width and height, and will default the pixel aspect ratio to 1, but Python doesn't. I've added in the extra bindings for that. (plus a test to confirm)

HughMacdonald commented:
I've rebased with master, and addressed the latest notes.

I'm off for the long weekend now (going to Rome!), so will look at any more comments on Tuesday.

Have a good weekend!

johnhaddon commented:
Grazie mille! I've made a couple of final tweaks and merged. Thanks for all your hard work on this one - although it's not the most exciting work, hopefully it's provided a useful tour of the various corners of GafferImage. Have a great long weekend...

HughMacdonald commented:
I've just realised that this PR was merged with code that would change the format for any script from before 0.17.0.0. However, this PR was merged into 0.16.0.2 (or was it 0.16.0.1?), so this would need to be changed, right? Otherwise scripts saved with 0.16.0.2 would have exclusive image bounds, but would also trigger the format compatibility code.

andrewkaufman commented:
Actually no, it is merged on master, but 0.16.0.* were all tagged of off 0.16.0.0 directly instead of master, so we could patch a couple bugs here at IE without introducing much new code. So 0.17.0.0 will be the first version with this merged. The issue you're noting is currently affecting personal builds off of master (since the SConstruct version isn't bumped yet), but I've been willing to live with that for the time being. We'll hopefully get 0.17 out soon (maybe after your Writer changes).

johnhaddon commented:
I think we could have handled the maintenance releases a bit better to result in less confusion. Here's what I think our standard way of doing this should be - not sure if there was a reason we didn't do that this last couple of times?

- When we need to make a maintenance release, branch off 0.X_maintenance from the tag for 0.X.
- Cherry pick across fixes from master to 0.X_maintenance as appropriate.
- Bump the minor/patch version on 0.X_maintenance as appropriate (leave the version alone on master).
- Tag 0.X.minor.patch off of 0.X_maintenance.
- Release from the tag.
- Repeat as necessary, continuing along on the 0.X_maintenance branch.

HughMacdonald commented:
Ahh - great - I wasn't aware of how this worked. I'd seen the version ticking up in the master branch, and had thought that it was that that was increasing.

I'm just wondering whether, once a version has been released, is it worth putting the master branch into what will be the next release, in a 'dev' mode. I can see compatibility code having issues in this kind of situation where the master branch (under dev for, say, 0.17.0.0) is thinking that it's 0.16.0.3, whereas the code is presuming that it's running under 0.17.0.0 (or higher)
